[GEOT-6682] WFS 2.0 GetFeature reponse parsing fails on "boundedBy"

awaterme opened the pull request:
Pull request for [GEOT-6682](https://osgeo-org.atlassian.net/browse/GEOT-6682). 

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/master/CONTRIBUTING.md) 
- [ ] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [x] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer (there is an automatic PR check verifying this, check this when it turns green).

The following are required only for core and extension modules (they are welcomed, but not required, for unsupported modules):
- [x] There is an issue in [Jira](https://osgeo-org.atlassian.net/projects/GEOT) describing the bug/task/new feature (a notable exemptions is, changes not visible to end users). The ticket is for the GeoTools project, if the issue was found elsewhere it's a good practice to link to the origin ticket/issue.
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message(s) must be in the form "[GEOT-XYZW] Title of the Jira ticket"
- [x] New unit tests have been added covering the changes
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by travis-ci after opening this PR) _I am not sure, but hope so_
- [ ] Documentation has been updated accordingly. _Not required in this case_

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

jodygarnett commented on modules/ogc/net.opengis.wfs/src/net/opengis/wfs20/impl/SimpleFeatureCollectionTypeImpl.java:
```diff
@@ -178,7 +178,7 @@ public Object eGet(int featureID, boolean resolve, boolean coreType) {
     switch (featureID) {
       case Wfs20Package.SIMPLE_FEATURE_COLLECTION_TYPE__BOUNDED_BY:
-        setBoundedBy((EnvelopePropertyType)newValue);
+        setBoundedBy(newValue instanceof EnvelopePropertyType ? (EnvelopePropertyType) newValue : null);
```
Q: What is it getting back that is not an EnvelopePropertyType?

jodygarnett replied:
I see, it is explicitly allowed to be null:

```
<complexType name="BoundingShapeType">
  <sequence>
    <choice>
      <element ref="gml:Envelope"/>
      <element ref="gml:Null"/>
    </choice>
  </sequence>
  <attribute name="nilReason" type="gml:NilReasonType"/>
</complexType>
```

awaterme replied:
> 
> 
> Q: What is it getting back that is not an EnvelopePropertyType?

Hi Jody, it get an empty String, because the parsing failed.

awaterme replied:
> 
> 
> I see, it is explicitly allowed to be null:
> 
> ```
> <complexType name="BoundingShapeType">
>   <sequence>
>     <choice>
>       <element ref="gml:Envelope"/>
>       <element ref="gml:Null"/>
>     </choice>
>   </sequence>
>   <attribute name="nilReason" type="gml:NilReasonType"/>
> </complexType>
> ```

When I look into the actual schema (http://schemas.opengis.net/wfs/2.0/wfs.xsd) I see the following:

![grafik](https://user-images.githubusercontent.com/11347841/91268715-1a341700-e776-11ea-9eb2-77b076744029.png)

Are GeoTools using this schema?

I think the problem is the "any" type...

aaime replied:
GeoTools is using that schema. Being a schema driven parser, it might well be that its having problems when dealing with a "any" target, it probably does not know what bindings to use. 
@awaterme if you're interested, try adding a breakpoint in XSAnyTypeBinding and see if that's the one it is using to parse stuff.
@nmco and @fernandor777 deal with this code more than I do, maybe they have a more educated guess as to what is happening.

awaterme commented:
I have adjusted the solution. I hope it is acceptable now: 

* I have rolled back the modifications of the generated code
* I have adjusted the bindings instead. The ClassCastException is resolved. The prior GT implementation always had a "null" bounding box. Now the bounding box is nested into the FeatureMap. The double nesting is the only solution I found which is accepted by the generated EMF model. 

I would be glad if the fix could make it into the GeoTools 24 release, although it is a little late already.

 BTW: The `XSAnyTypeBinding` is not executed here, because the nested element is an GML Envelope. The binding for the envelope is used instead.

aaime commented on modules/extension/xsd/xsd-wfs/src/main/java/org/geotools/wfs/v2_0/bindings/EnvelopePropertyTypeBinding.java:
```diff
@@ -19,20 +19,41 @@
 
 public class EnvelopePropertyTypeBinding extends AbstractComplexBinding {
 
+    private Class<ReferencedEnvelope> referencedEnvelopeClass = ReferencedEnvelope.class;
```
All of these... they look like they should be constants, no?

awaterme replied:
Hi Andrea,

could be constants, but in that case on class loading of EnvelopePropertyTypeBinding the EMF Factories would also be bootstrapped. In my experience that sometimes causes trouble, maybe in testing scenarios, etc.
On the other hand I was not sure about the lifespan of the EnvelopePropertyTypeBinding. If it is created often, statics might make sense...
As you like. Let me know what you prefer please.

Best regards,
Andreas

aaime replied:
If memory serves me right, bindings are instantiated anew every time we parse or encode something. Each parse/encode gets its own dedicated Pico Container context, bindings are registered in there, every time. I'm not sure if they are just registered as classes, and then instantiated every time they are needed, or not. If it was the former, that would mean they get instantiated multiple times per parse/encode operation, too.

awaterme replied:
Do you prefer static fields then?

aaime replied:
Personally I do, yes.

awaterme commented:
Hi Andrea,

next attempt with static fields...

Best regards,
Andreas

aaime commented:
Think it looks good. Constants are not named after Java coding conventions, should look like "THIS_IS_A_CONSTANT", but they are private, so whatever... guess I'll have to add a Checkstyle check for it, later down the road, so that it cannot happen any longer.
In the meantime, squash merging.
